src/dvc_data/index/checkout.py

```diff
@@ -318,13 +314,45 @@ def meta_cmp_key(meta):
                 _add_create(change.new)
 
             elif old_isexec != new_isexec and not new_isdir:
-                files_chmod.append(change.new)
+                ret.files_chmod.append(change.new)
             else:
                 continue
         else:
             raise AssertionError()
 
-        changes[change.typ].append(change)
+        ret.changes[change.typ].append(change)
+
+    return ret
+
+
+def checkout(
+    index: "BaseDataIndex",
+    path: str,
+    fs: "FileSystem",
+    old: Optional["BaseDataIndex"] = None,
+    delete: bool = False,
+    callback: "Callback" = DEFAULT_CALLBACK,
+    latest_only: bool = True,
+    update_meta: bool = True,
+    jobs: Optional[int] = None,
+    storage: str = "cache",
+    prompt: Optional[Callable] = None,
+    relink: bool = False,
+    force: bool = False,
+    allow_missing: bool = False,
+    state: Optional["StateBase"] = None,
+    **kwargs,
+) -> Dict[str, List["Change"]]:
+
+    failed = []
+
+    if callback == DEFAULT_CALLBACK:
+        cb = callback
+    else:
+        cb = Callback.as_tqdm_callback(desc="Comparing indexes", unit="entry")
```

Comment:
This pattern is faulty, will take another look.

Reply:
Diffing is an internal implementation detail. You don’t expect `callback` to be used for anything except for keeping track of actual checkouts.

A separate optional callback like `diff_callback` might be better.

That, or `checkout` should not use `fsspec.callbacks` and should provide something of its own.

Reply:
This may also point to a problem why checkout takes an index instead of Diff giving control to the caller instead?

Reply:
> This may also point to a problem why checkout takes an index instead of Diff giving control to the caller instead?

Agreed! We should have something like `prepare_fetch` or `collect` or something to do the diffing separately.